File: implementation_generated_by_AI/Chapter_12_Binary_Search_Tree/Problems/12-2 Radix trees/12-2.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <fstream>
#include <algorithm>

class TrieNode {
public:
    std::map<char, TrieNode*> children;
    bool is_end_of_string;

    TrieNode() : is_end_of_string(false) {}
};

class RadixTree {
private:
    TrieNode* root;

    void pre_order_traversal(TrieNode* node, const std::string& prefix, std::vector<std::string>& result) {
        if (node->is_end_of_string) {
            result.push_back(prefix);
        }
        /*
        In C++, std::map inherently maintains its elements in ascending order by key. 
        Therefore, you don't need to sort the std::map when iterating over it; it will automatically be in ascending order. 
        */
        for (const auto& pair : node->children) {
            pre_order_traversal(pair.second, prefix + pair.first, result);
        }
    }

    void add_edges(TrieNode* node, const std::string& prefix, std::ofstream& dot) {
        for (const auto& pair : node->children) {
            TrieNode* child = pair.second;
            std::string child_prefix = prefix + pair.first;
            // add double quotes in case "prefix" is an empty string
            dot << "    \"" << prefix << "\" -> \"" << child_prefix << "\" [label=" << pair.first << "];\n";
            if (child->is_end_of_string) {
                dot << "    " << child_prefix << " [style=filled, fillcolor=lightgrey];\n";
            }
            add_edges(child, child_prefix, dot);
        }
    }

public:
    RadixTree() {
        root = new TrieNode();
    }

    void insert(const std::string& string) {
        TrieNode* node = root;
        for (char charac : string) {
            if (node->children.find(charac) == node->children.end()) {
                node->children[charac] = new TrieNode();
            }
            node = node->children[charac];
        }
        node->is_end_of_string = true;
    }

    std::vector<std::string> lexicographic_sort() {
        std::vector<std::string> result;
        pre_order_traversal(root, "", result);
        return result;
    }

    void visualize(const std::string& filename) {
        std::ofstream dot(filename);
        dot << "digraph RadixTree {\n";
        dot << "    node [shape=circle];\n";
        add_edges(root, "", dot);
        dot << "}\n";
        dot.close();

        std::string command = "dot -Tpng " + filename + " -o " + filename + ".png";
        system(command.c_str());

        // display the image
        command = "xdg-open " + filename + ".png";
        system(command.c_str());
    }
};

// Example usage
int main() {
    std::vector<std::string> bit_strings = {"1011", "10", "011", "100", "0"};
    RadixTree radix_tree;

    for (const auto& bit_string : bit_strings) {
        radix_tree.insert(bit_string);
    }

    std::vector<std::string> sorted_strings = radix_tree.lexicographic_sort();
    for (const auto& str : sorted_strings) {
        std::cout << str << ", ";
    }
    std::cout << std::endl;

    // Visualize the radix tree
    radix_tree.visualize("./graphviz/radix_tree.dot");

    return 0;
}

File: implementation_generated_by_AI/Chapter_12_Binary_Search_Tree/Problems/12-2 Radix trees/12-2.py
from typing import List, Dict
from graphviz import Digraph # type: ignore

class TrieNode:
    def __init__(self) -> None:
        self.children: Dict[str, TrieNode] = {}
        self.is_end_of_string: bool = False

class RadixTree:
    def __init__(self) -> None:
        self.root: TrieNode = TrieNode()

    def insert(self, string: str) -> None:
        node: TrieNode = self.root
        for char in string:
            if char not in node.children:
                node.children[char] = TrieNode()
            node = node.children[char]
        node.is_end_of_string = True

    def lexicographic_sort(self) -> List[str]:
        result: List[str] = []
        self._pre_order_traversal(self.root, "", result)
        return result

    def _pre_order_traversal(self, node: TrieNode, prefix: str, result: List[str]) -> None:
        if node.is_end_of_string:
            result.append(prefix)
        for char in sorted(node.children.keys()):
            self._pre_order_traversal(node.children[char], prefix + char, result)

    def visualize(self, filename: str) -> None:
        """
        Generate a visual representation of the radix tree using graphviz.
        """
        dot = Digraph(comment='Radix Tree')
        dot.attr('node', shape='circle')
        self._add_edges(self.root, "", dot)
        
        # Configure the graph to be laid out from top to bottom
        dot.attr(rankdir='TB')
        dot.render(filename, format='png', view=True)

    def _add_edges(self, node: TrieNode, prefix: str, dot: Digraph) -> None:
        # for char, child in node.children.items():
        for char in sorted(node.children.keys()):
            child = node.children[char]
            child_prefix = prefix + char
            dot.node(child_prefix, label=child_prefix)
            dot.edge(prefix, child_prefix, label=char)
            self._add_edges(child, child_prefix, dot)
            if child.is_end_of_string:
                dot.node(child_prefix, label=child_prefix, style='filled', fillcolor='lightgrey')


# Example usage
if __name__ == "__main__":
    bit_strings: List[str] = ["1011", "10", "011", "100", "0"]
    radix_tree: RadixTree = RadixTree()
    for bit_string in bit_strings:
        radix_tree.insert(bit_string)

    sorted_strings: List[str] = radix_tree.lexicographic_sort()
    print(sorted_strings)  # Output: ['0', '011', '10', '100', '1011']

    # Visualize the radix tree
    radix_tree.visualize('./graphviz/radix_tree.dot')
    